sklearn/feature_selection/from_model.py

```diff
@@ -123,25 +129,49 @@ class SelectFromModel(BaseEstimator, SelectorMixin, MetaEstimatorMixin):
     threshold_ : float
         The threshold value used for feature selection.
     """
-    def __init__(self, estimator, threshold=None, prefit=False, norm_order=1):
+    def __init__(self, estimator, threshold=None, prefit=False,
+                 max_features=None, norm_order=1):
         self.estimator = estimator
         self.threshold = threshold
         self.prefit = prefit
+        self.max_features = max_features
         self.norm_order = norm_order
 
+    def _check_params(self, X, y):
+        X, y = check_X_y(X, y)
+
+        if self.max_features is None:
+            return
+
+        if isinstance(self.max_features, numbers.Integral):
+            if 0 <= self.max_features <= X.shape[1]:
+                return
+
+        raise ValueError(
+            "max_features should be >=0 and <= n_features;"
+            " got %r." % self.max_features)
+
     def _get_support_mask(self):
         # SelectFromModel can directly call on transform.
         if self.prefit:
             estimator = self.estimator
         elif hasattr(self, 'estimator_'):
             estimator = self.estimator_
         else:
-            raise ValueError(
-                'Either fit SelectFromModel before transform or set "prefit='
-                'True" and pass a fitted estimator to the constructor.')
+            raise ValueError('Either fit the model before transform or set'
+                             ' "prefit=True" while passing the fitted'
+                             ' estimator to the constructor.')
         scores = _get_feature_importances(estimator, self.norm_order)
         threshold = _calculate_threshold(estimator, scores, self.threshold)
-        return scores >= threshold
+        mask = np.zeros_like(scores, dtype=bool)
+        if self.max_features is not None:
+            candidate_indices = \
+                np.argsort(-scores, kind='mergesort')[:self.max_features]
+            mask[candidate_indices] = True
+        else:
+            mask = np.logical_not(mask)
```

Comment:
Nitpick: I'd rather see this as ones_like, with zeros_like in the if case. But whatever.

Reply:
I approve this change :)